DnfContext: make dnf_context_module_install like `dnf module install`

jlebon opened the pull request:
This significantly revamps `dnf_context_module_install`. It fixes some
issues noticed during review, but also aligns the logic closer to what
`dnf module install` does.

jlebon committed:
DnfContext: make dnf_context_module_install like `dnf module install`

This significantly revamps `dnf_context_module_install`. It fixes some
issues noticed during review, but also aligns the logic closer to what
`dnf module install` does.

jlebon committed:
DnfContext: improve error-reporting for modular functions

As a library, we need to be able to report errors directly to the
application and not log it ourselves. Applications should take care of
displaying the errors to users.

Additionally, let's not log critical errors for things which are due to
user-input, such as a non-existent module name. In rpm-ostree for
example, we run enable `G_DEBUG=fatal-warnings` in some test scenarios
to ensure that we don't trigger any warnings. This then causes
rpm-ostree to crash if doing `rpm-ostree module install badname`.

This patch changes things so that instead of logging errors, we
concatenate them together and return them through the GError API.

= changelog =
msg: Improve error-reporting for modular functions
type: bugfix

jlebon commented:
I can split this further into separate PRs if desired.

j-mracek commented on libdnf/dnf-context.cpp:
```diff
@@ -3481,12 +3483,14 @@ dnf_context_module_install(DnfContext * context, const char ** module_specs, GEr
                         } else {
                             profiles.push_back(modpkg->getDefaultProfile());
                         }
+                        auto nsvca = modpkg->getFullIdentifier();
```
I suggest that there is an issue in workflow. Before we can look at profiles modular sack must be resolved. If I am not mistaken call `auto solver_error = recompute_modular_filtering(container, sack, hotfixRepos);` is to late. Then I should take only the latest active module to collect profiles for install.

j-mracek replied:
There is also a problem that modules or NSVCA is not installable. Only profiles that contains names of packages. can be installed.

jlebon replied:
Thanks for the feedback. So in other words roughly what you're suggesting is this order:
- `modify_module_dict_and_enable_stream()`
- `recompute_modular_filtering()`
- `container->install()` profiles we want
- `dnf_context_install()` for artifacts of each profile

?

But doesn't `modify_module_dict_and_enable_stream()` already activate the right modules from which we should be searching for profiles? Testing this code with https://github.com/coreos/rpm-ostree/pull/2760, it seems to work well. Though I haven't tested all the corner cases.

> There is also a problem that modules or NSVCA is not installable. Only profiles that contains names of packages. can be installed.

Yes, we're only installing at the profile level here. This is following the same semantics as `dnf module install`, where it auto-fills the stream (if there is only one) and the profile (if there is a default one).

j-mracek replied:
I guess that there will be a difference when modules in the same stream will have a different set of profiles or profiles will have a different set of rpms names. I can provide a real examples if you require it.

I suggest that we have to resolve the workflow issue first and then we can look for a solution that was original target of PR. Do you agree?

jlebon replied:
Hi @j-mracek, thanks for the feedback. I reworked things now to make this more or less in line with what `dnf module install` does. Consequently, we now correctly depsolve the modular sack before looking at profiles and packages.

jlebon committed:
ModulePackageContainer: factor out helper to get latest module

Prep for next patch.

jlebon committed:
dnf-context: add some function comments

A few things I wrote down to help find my way around the modularity
code.

jlebon commented on libdnf/dnf-context.cpp:
```diff
@@ -3525,18 +3487,132 @@ dnf_context_module_install(DnfContext * context, const char ** module_specs, GEr
+    g_autoptr(GPtrArray) resolved = g_ptr_array_new_with_free_func (g_free);
+    for (auto it = std::make_move_iterator(all_resolved_module_dicts.begin());
+              it != std::make_move_iterator(all_resolved_module_dicts.end()); it++) {
+        auto module_spec = std::get<0>(*it);
+        auto nsvcap_obj = std::get<1>(*it);
+        auto module_dict = std::get<2>(*it);
```
I tried for a while to get `std::tie` to work here, but it simply doesn't like moving the Nsvcap object out of the tuple. We can look at this in a follow-up perhaps.

jlebon commented:
If the overall approach looks sane, I can work on adding tests.

jlebon commented:
Hmm, I'm trying to understand from the integration tests logs what part failed, but the output is quite large. In the tail of the output, I see:

```
2021-05-07T20:07:01.3167811Z 1 feature passed, 0 failed, 0 skipped
2021-05-07T20:07:01.3168378Z 32 scenarios passed, 0 failed, 0 skipped
2021-05-07T20:07:01.3168975Z 316 steps passed, 0 failed, 0 skipped, 0 undefined
2021-05-07T20:07:01.3169466Z Took 0m46.922s
2021-05-07T20:07:01.3179164Z ##[error]Process completed with exit code 1.
```

And quickly grepping the results from the other behave tests, they all say `0 failed`.

lukash commented:
@jlebon it's an occasional "infra" failure (of the overlay FS), you can search the logs for "OCI" to find it. Sorry about the inconvenience, it's hitting us quite often and I just can't seem to figure out what to do about it. I'll try to report to podman, but I don't have a reproducer outside of GitHub Actions. We could try to switch to docker...

Also @martinpitt would you have any idea?

I'll re-run the jobs for you (you can do it by clicking on the button in top right corner yourself), the failing action for future reference is here: https://github.com/rpm-software-management/libdnf/pull/1207/checks?check_run_id=2530538673

martinpitt commented:
The [re-run failed again](https://github.com/rpm-software-management/libdnf/pull/1207/checks?check_run_id=2546815978#step:4:17878) with

     Error: executable file `behave` not found in $PATH: No such file or directory: OCI not found

I'm afraid I have no idea what `behave` is supposed to be. I [see](https://github.com/rpm-software-management/libdnf/actions/runs/821343443/workflow#L45) that you are already using `options: --privileged`, which ought to circumvent the [faccessat() dealbreaker](https://bugzilla.redhat.com/show_bug.cgi?id=1900021). You can still try `--security-opt=seccomp=unconfined` just in case that helps. Other than that, I'm afraid I have no idea :cry:

jlebon commented:
Ahh yup, OK that error is a recent podman regression I think: https://github.com/containers/podman/issues/10293 (I hit the same thing locally with my pet container after updating to https://bodhi.fedoraproject.org/updates/FEDORA-2021-fc15685f73).

lukash commented:
@martinpitt `behave` is a test framework, in this context it's just an executable on the filesystem, I've seen the same error happen with `chown` and others. Thank you, I just thought I'd try to ask :slightly_smiling_face: 

@jlebon I'm quite sure it's not https://github.com/containers/podman/issues/10293, apparently you can get the "OCI not found" error message in multiple various cases, I've observed this (almost) from the beginning of using GitHub Actions, but somehow the frequency seems to be increasing. It's non-deterministic, sometimes it works, sometimes it happens once in the hundreds of `behave` invocations, sometimes even more.

Re-running again, off to attempt a podman bug report.

lukash commented:
Reported here: https://github.com/containers/podman/issues/10321

jlebon commented:
Thanks @lukash, CI looks green now! :)

@j-mracek Could you have a look at this now? Happy to fix things up as necessary to drive this forward. Also open to have a higher-bandwidth conversation over video if you'd like.

Thanks!

j-mracek commented:
@jlebon I will take a look in next days.

j-mracek commented on libdnf/module/ModulePackageContainer.cpp:
```diff
@@ -851,6 +851,24 @@ ModulePackageContainer::getModuleState(const std::string& name)
+{
+    ModulePackage * latest = nullptr;
+    for (ModulePackage * module: modulePackages) {
+        if (isModuleActive(module->getId()) || !activeOnly) {
```
I would recommend to test `!activeOnly` first.
`if (!activeOnly || isModuleActive(module)) {`

j-mracek commented on libdnf/dnf-context.cpp:
```diff
@@ -3480,6 +3480,9 @@ dnf_context_module_install(DnfContext * context, const char ** module_specs, GEr
                         if (resolved_spec.first->getProfile() != "") {
                             profiles = modpkg->getProfiles(resolved_spec.first->getProfile());
+                            if (profiles.empty()) {
+                                throw std::runtime_error("No profile found matching " + resolved_spec.first->getProfile());
```
Please use directly gerror.

jlebon replied:
A gerror would mean returning immediately, but I'm trying to match dnf's (and libdnf's other modular functions) behaviour of continuing to process all the inputs before returning errors.

j-mracek replied:
I am sorry but there must be some misunderstanding. When you throw, the function stops immediately. But the exception is directly catch by `} CATCH_TO_GERROR(NULL)` and transformed into g_error. Then I suggest to set g_error directly and return.

jlebon replied:
Ahh yes good catch! What I meant was for the exception to be caught by the inner `try`/`catch`, so that it just adds to the `messages` vector, but in copy/pasting I forgot to generalize it to `catch (const std::exception & ex)` instead of just a specific error. I'll change that then. Does that make sense?

j-mracek commented on libdnf/dnf-context.h:
```diff
@@ -311,11 +311,12 @@ gboolean         dnf_context_module_enable              (DnfContext * context,
  * Since: 0.63.0
  **/
-gboolean         dnf_context_module_install              (DnfContext * context,
+char**           dnf_context_module_install              (DnfContext * context,
```
Be honest - such a change is always problematic. It requires change in so name version or provide the change as a new function to keep compatibility. It also makes a difference to other functions that handles modularity.

j-mracek replied:
May be the output from `getInstalledProfiles();` can be sufficient for your user case.

jlebon replied:
> Be honest - such a change is always problematic. It requires change in so name version or provide the change as a new function to keep compatibility. It also makes a difference to other functions that handles modularity.

When I initially opened this PR, there were no releases yet with the new API so it was safe to change it. But with the API out in v0.63.0, that's no longer the case. So agreed, this is not safe to do.

I'll add a separate API for this.

j-mracek replied:
@jlebon The most important questions is: Do you really need it? Probably there is a way how you can get what do you need without the change of API. Please could you explain me, why do you need it? Or how exactly you will use the returned strings?

jlebon replied:
I'd really like to include the fully resolved NSVCA/P in the OSTree metadata just on principle so that it's inspectable. Otherwise, because modules purely live in yumrepo metadata and not e.g. on disk, there isn't really any trace of what exact modules and their versions are part of an OSTree commit.

But let me see, I think we might be able to do without it and just store the user-provided string as is. The biggest concern there is handling globby patterns like `postgresq?:13/client` (obviously not a realistic example), but maybe better anyway to just deny globby patterns and force users to type out the module spec more explicitly.

j-mracek replied:
`postgresq?:13/client` is nice and easy to hadle example, but `nodejs:*` is more interesting example. <spec> matches multiple streams.

jlebon replied:
OK, dropped the need for this now and just went with the naive approach of storing the user-provided string as is. At least the modular RPMs are still in the rpmdb, so one can sort of infer what modules were installed. We'll look at adding this in a follow-up if we really need it.

j-mracek commented on libdnf/dnf-context.cpp:
```diff
@@ -3525,18 +3487,132 @@ dnf_context_module_install(DnfContext * context, const char ** module_specs, GEr
+                    }
+
+                    if (active_modules.empty()) {
+                        throw std::runtime_error(tfm::format(_("No active module packages found for module spec '%s'"), module_spec));
```
Please, setup directly g_error

j-mracek commented on libdnf/dnf-context.cpp:
```diff
@@ -3525,18 +3487,132 @@ dnf_context_module_install(DnfContext * context, const char ** module_specs, GEr
+                        if (dnf_repo_get_module_hotfixes(repo)) {
+                            hotfixRepos.push_back(dnf_repo_get_id(repo));
+                        }
+                    }
```
`hotfixRepos` is already defined and filled above (3475)

j-mracek commented on libdnf/dnf-context.cpp:
```diff
@@ -3525,18 +3487,132 @@ dnf_context_module_install(DnfContext * context, const char ** module_specs, GEr
+                            query = hy_query_clone(base_nosrc_query);
+                            hy_query_filter(query, HY_PKG_NAME, HY_EQ, pkgname);
+                            if (hy_query_is_empty(query)) {
+                                throw std::runtime_error(tfm::format(_("No match for package '%s' for module spec %s"), pkgname, module_spec));
```
Please set g_error directly

j-mracek commented:
I also suggest that module profile installed by C API will be impossible to remove by `dnf module remove` command because packages reason in C API is user, but expected is group.

jlebon commented:
Updated! I squashed some of the commits together to make review easier. I didn't rebase it so you can use the inter-diff to see what changed.

> I also suggest that module profile installed by C API will be impossible to remove by `dnf module remove` command because packages reason in C API is user, but expected is group.

Can you expand on this? I think I understand what you mean, but I couldn't find in the code what marks a module's install reason.

Worth noting that in rpm-ostree at least, we always re-assemble the rootfs from scratch, so there isn't a need to remove modules the same way: we just stop adding it in the new rootfs (we do support removing base packages, but to do that at the modular level would first require fixing https://github.com/rpm-software-management/libdnf/pull/1207#discussion_r645105546; until then, just removing the modular RPMs directly should work fine). Obviously, I do want this code to be leveraged by `dnf module install` too in the future though, so happy to fix up things as needed here to help with that!

j-mracek commented:
@jlebon I suggest that the PR looks good to me. It enhance behaviour of `dnf_context_module_install`. I propose fix the problem with `dnf module remove` in next PR. Is there anything else what you want to include in this PR (tests?)?

> > I also suggest that module profile installed by C API will be impossible to remove by `dnf module remove` command because packages reason in C API is user, but expected is group.
> 
> Can you expand on this? I think I understand what you mean, but I couldn't find in the code what marks a module's install reason.
> 

The difference is that dnf module install commands shares implementation with `dnf group install`. The difference is that after installation of packages installed from profiles (not their dependencies) they are marked with reason group. Normally they are installed as user.
Because the context part in libdnf does not support groups, there is not yet implemented the mechanism to mark installed packages with reason group.

To investigate stored reasons you can use `dnf repoquery --installed --qf "%{name}.%{arch} %{reason}"`

jlebon committed:
dnf-context: fix iterator copy vs reference

We were copying the map and modifying the copy rather than the original.
This caused issues in callers which assume that the argument passed was
modified so that only a single stream was retained.

jlebon commented:
> @jlebon I suggest that the PR looks good to me. It enhance behaviour of `dnf_context_module_install`. I propose fix the problem with `dnf module remove` in next PR. Is there anything else what you want to include in this PR (tests?)?

Thanks, I've added some tests to this now so this is ready to go! And doing that uncovered a bug. :tada: Split that out as a separate commit.
